xrdp/xrdp_mm.c

```diff
@@ -1192,24 +1094,233 @@ dynamic_monitor_data(intptr_t id, int chan_id, char *data, int bytes)
+            break;
+        case WMRZ_ENCODER_CREATE:
+            if (mm->encoder == NULL)
+            {
+                mm->encoder = xrdp_encoder_create(mm);
+            }
+            advance_resize_state_machine(mm->resize_state_machine, description, WMRZ_SERVER_INVALIDATE);
+            break;
+        case WMRZ_SERVER_INVALIDATE:
+            /* redraw */
+            error = xrdp_bitmap_invalidate(wm->screen, 0);
+            if (error != 0)
+            {
+                LOG(LOG_LEVEL_INFO, "process_dynamic_monitor_description: xrdp_bitmap_invalidate failed %d", error);
+                return advance_error(error, mm->resize_state_machine, description);
+            }
+            if (module != 0)
+            {
+                // Ack all frames to speed up resize.
+                module->mod_frame_ack(module, 0, INT_MAX);
+            }
+            advance_resize_state_machine(mm->resize_state_machine, description, WMRZ_COMPLETE);
+            break;
+        default:
+            break;
+    }
+    return 0;
+}
+
+/******************************************************************************/
+static int
+dynamic_monitor_process_queue(struct xrdp_mm *self)
+{
+    struct dynamic_monitor_description *description;
+
+    LOG_DEVEL(LOG_LEVEL_TRACE, "dynamic_monitor_process_queue:");
+
+    if (self == 0)
+    {
+        return 0;
+    }
+    if (self->wm->login_state != WMLS_CLEANUP && self->wm->login_state != WMLS_INACTIVE)
+    {
+        LOG(LOG_LEVEL_INFO, "dynamic_monitor_process_queue: Not trying to resize while logging in! State was %s", wm_login_state_to_str(self->wm->login_state));
+        return 0;
+    }
+    if (self->resize_queue == 0)
+    {
+        // Potentially the client does not support resizing, so the queue was never created.
+        return 0;
+    }
+    if (self->resize_queue->count <= 0)
+    {
+        return 0;
+    }
+
+    while (self->resize_queue->count > 0)
+    {
+        description = (struct dynamic_monitor_description *)list_get_item(self->resize_queue, 0);
+        if (description->state == WMRZ_COMPLETE)
+        {
+            LOG(LOG_LEVEL_INFO, "dynamic_monitor_process_queue: Clearing completed resize (w: %d x h: %d). It took %d MS.", description->description.session_width, description->description.session_height, g_time3() - description->start_time);
+        }
+        else if (description->state == WMRZ_ERROR)
+        {
+            LOG(LOG_LEVEL_INFO, "dynamic_monitor_process_queue: Clearing completed resize (w: %d x h: %d)", description->description.session_width, description->description.session_height);
+        }
+        else if (description->description.session_width <= 0 && description->description.session_height <= 0)
+        {
+            LOG(LOG_LEVEL_INFO, "dynamic_monitor_process_queue: Not allowing resize due to invalid dimensions (w: %d x h: %d)", description->description.session_width, description->description.session_height);
```

Comment:
Why is this condition of invalid bounds handled in both this function and the process_dynamic_monitor_description function? Can we simplify this by handling this condition only in process_dynamic_monitor_description and seeing the state to WMRZ_ERROR?

Reply:
Great catch! This code has been in the works for a while so I missed that :)